### PR TITLE
[SETUP][BOOTDATA][MEDIA][NTGDI] Enable MS Sans Serif

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-16166](https://jira.reactos.org/browse/CORE-16166), [CORE-15559](https://jira.reactos.org/browse/CORE-15559)

- Enable "MS Sans Serif" fonts.
- Modify the initial font substitutes.
- Fix `freetype.c` a little.

## TODO
- [ ] Fix metrics.
- [ ] Fix StrikeOut.

### CODE REVIEW

**Comment by katahiromz** [comment]
In startup of AbiWord:
```txt
(win32ss/gdi/ntgdi/freetype.c:4693) Can't find OS/2 table - not TT font?
(win32ss/gdi/ntgdi/freetype.c:4700) Can't find HHEA table - not TT font?
(win32ss/gdi/ntgdi/freetype.c:7176) WARNING: Could not find desired charmap!

*** Fatal System Error: 0x00000001
                       (0x000000A2,0x00000000,0x0000FFFF,0x00000000)

[7h
Entered debugger on embedded INT3 at 0x0008:0x809565c4.
kdb:> bt
Eip:
<ntoskrnl.exe:1565c4 (sdk/lib/rtl/i386/debug_asm.S:56 (RtlpBreakWithStatusInstruction))>
Frames:
<ntoskrnl.exe:8a53b (ntoskrnl/ke/bug.c:1136 (KeBugCheckWithTf))>
Couldn't access memory at 0xF5CB3CF4!
kdb:> 
```

**Comment by katahiromz** [comment]
Loaded:
loaded

**Comment by katahiromz** [comment]
FontPalette13-after
Able to render.

**Comment by katahiromz** [comment]
CORE-15559-fixed
CORE-15559 is fixed.

**Comment by katahiromz** [comment]
Russian rendering is strange.
russian-strange

**Comment by katahiromz** [comment]
metrics-broken
Metrics are broken.

**Comment by Doug-Lyons** [comment]
Please look at readme.abw in Abiword as shown here:
Abiword_Broken

**Comment by katahiromz** [comment]
```txt
(win32ss/gdi/ntgdi/freetype.c:691) Fonts registry is empty.
```
???

**Comment by HBelusca** [comment]
> In startup of AbiWord:
> 
> ```
> (win32ss/gdi/ntgdi/freetype.c:4693) Can't find OS/2 table - not TT font?
> (win32ss/gdi/ntgdi/freetype.c:4700) Can't find HHEA table - not TT font?
> (win32ss/gdi/ntgdi/freetype.c:7176) WARNING: Could not find desired charmap!
> 
> *** Fatal System Error: 0x00000001
>                        (0x000000A2,0x00000000,0x0000FFFF,0x00000000)
> 
> �[7h�
> Entered debugger on embedded INT3 at 0x0008:0x809565c4.
> kdb:> bt
> Eip:
> <ntoskrnl.exe:1565c4 (sdk/lib/rtl/i386/debug_asm.S:56 (RtlpBreakWithStatusInstruction))>
> Frames:
> <ntoskrnl.exe:8a53b (ntoskrnl/ke/bug.c:1136 (KeBugCheckWithTf))>
> Couldn't access memory at 0xF5CB3CF4!
> kdb:> 
> ```

Why did this ever happen?

**Comment by SergeGautherie** [comment]
'Thai' name reports '8' twice?

**Comment by katahiromz** [comment]
Thai font has 8px regular and 8px bold.

File: base/setup/lib/muifonts.h [hunk 1]
```diff
@@ -388,7 +384,7 @@ MUI_SUBFONT HindiFonts[] =
     { L"Courier New Greek,161",  L"Courier New,161" },
     { L"Courier New TUR,162",    L"Courier New,162" },
     { L"Fixedsys",               L"Fixedsys Excelsior 3.01-L2" },
-    { L"Helv",                   L"Tahoma" },
+    { L"Helv",                   L"MS Sans Serif" },
     { L"Helvetica",              L"Arial" },
     { L"MS Sans Serif",          L"FreeSans" },
```
Comment (SergeGautherie) [inline base/setup/lib/muifonts.h]: Double-checking: what about this substitute? (Helv: Tahoma vs FreeSans...)

Comment (katahiromz) [reply base/setup/lib/muifonts.h]: Helv is Win16-compatible font substitute that is mapped to MS Sans Serif in Windows. Some languages are lacking some language-specific glyphs in MS Sans Serif.

File: media/inf/font.inf [hunk 2]
```diff
@@ -50,7 +48,7 @@ HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Tms Rmn",0x
 [Font.Greek.Reg]
 HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Courier",0x00000000,"FreeMono"
 HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Fixedsys",0x00000000,"Fixedsys Excelsior 3.01-L2"
-HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Helv",0x00000000,"DejaVu Sans"
+HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Helv",0x00000000,"MS Sans Serif"
 HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","Helvetica",0x00000000,"Arial"
 HKLM,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\FontSubstitutes","MS Sans Serif",0x00000000,"DejaVu Sans"
```
Comment (SergeGautherie) [inline media/inf/font.inf]: Remove this one too?

**Comment by katahiromz** [comment]
I don't have any idea why the metrics are broken...

**Comment by SergeGautherie** [comment]
Is freetype.c fix ready to be committed separately?
As for MS Sans Serif, would committing this PR as is be an improvement or a regression, compared to Tahoma substitute?

**Comment by katahiromz** [comment]
No, it isn't. We can suppose freetype.c has bugs. MS Sans Serif is a raster font collection. It might fix CORE-15559.

**Comment by tkreuzer** [comment]
> 
> 
> In startup of AbiWord:
> 
> ```
> (win32ss/gdi/ntgdi/freetype.c:4693) Can't find OS/2 table - not TT font?
> (win32ss/gdi/ntgdi/freetype.c:4700) Can't find HHEA table - not TT font?
> (win32ss/gdi/ntgdi/freetype.c:7176) WARNING: Could not find desired charmap!
> 
> *** Fatal System Error: 0x00000001
>                        (0x000000A2,0x00000000,0x0000FFFF,0x00000000)
> 

This crash in win32k needs a fix in win32k. Even if there is something wrong with the font, it should not crash. The changes in win32k don't look like they would fix the crash, but maybe I'm missing something here. Can you provide more insight into that?

**Comment by Saibamen** [comment]
Please mark this Pull Request as Draft if is still WIP: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request

**Comment by katahiromz** [comment]
I lost interest for this issue. Somebody else can do. Closed.

**Pull request closed without merging**